### ISSUE TITLE
load model correction, SpineComputeROIs inconsistent error

### ISSUE THREAD

**Issue by fohofmann** [StanfordMIMI/Comp2Comp#35]
Hi Louis,

thank you for cleaning up :) Two issues I encountered with the latest version of your work:

# Downloading model from hugging face each run
The pipeline loads muscle/fat model from hugging face each run, regardless whether the model already exists in models/ or not.
> Downloading muscle/fat model from hugging face
100% [......................................................................] 138570720 / 138570720

Although this might drive your download rates at hugging face :), it's unnecessary 140mb traffic per segmentation. Thus, I changed *models.py* line [L71-72](https://github.com/StanfordMIMI/Comp2Comp/blob/464cb8dd3bc032090456b52589318aee0124d4f9/comp2comp/models/models.py#L71-72) to:
```python
try:
   filename = Models.find_model_weights(self.model_name, model_dir)
```

# Spine compute ROIs
In some dcm, the computation of the ROIs does not work.

> Finished SpineSegmentation with output keys dict_keys(['segmentation', 'medical_volume'])
Running SpineReorient with input keys odict_keys(['inference_pipeline', 'segmentation', 'medical_volume'])
Finished SpineReorient with output keys dict_keys([])
Running SpineComputeROIs with input keys odict_keys(['inference_pipeline'])
ERROR PROCESSING inputs/batch/27
Traceback (most recent call last):
  File "...Comp2Comp/bin/C2C", line 122, in spine_muscle_adipose_tissue pipeline()
  File "...Comp2Comp/comp2comp/inference_pipeline.py", line 48, in __call__  output = inference_class(self, **output)
  File "...Comp2Comp/comp2comp/spine/spine.py", line 143, in __call__ (spine_hus, rois, centroids_3d) = spine_utils.compute_rois(
  File "...Comp2Comp/comp2comp/spine/spine_utils.py", line 229, in compute_rois  two_largest = keep_two_largest_connected_components(slice)
  File "...Comp2Comp/comp2comp/spine/spine_utils.py", line 262, in keep_two_largest_connected_components mask[labels == sorted_indices[i] + 1] = 1
IndexError: index 1 is out of bounds for axis 0 with size 1

**The strange thing is, that this error does not occur consistently. Instead, if I rerun the pipeline using exactly the sam dcm, in sometimes everything works fine.** Any ideas where to start debugging? I assume that the problem is not `keep_two_largest_connected_components` itself but the result of `connectedComponentsWithStats`?!

Best, Felix

**Comment by louisblankemeier** [comment]
Hi Felix,

I have processed about 200 scans with the new Comp2Comp and have not run into this issue. 

To help you debug a bit, this function is responsible for selecting the 2 largest connected components within a sagittal plane that passes through the left/right center of a specific vertebra. These two connected components should correspond to the vertebral body and the spinous process. Perhaps there was a segmentation failure and you are only seeing 1 connected component on that sagittal plane? The way to check this would be to plot the input to keep_two_largest_connected_components(), which represents the segmentations on this sagittal plane. 

Please let me know if you figure out what is going on! Thank you!

**Comment by fohofmann** [comment]
Thank you Louis for your quick response. I also assume, it is some segmentation issue, maybe due to the "real-world" DICOMs (missing slices, larger spacing or sth else). I will come back to you, if I figured out the problem.
Just to my understanding: What is the reason, that c2c processes only series with more than 300 slices? I had to reduce this, as in our setting often thorax and (upper) abdomen are only available as separate series, both well below n=300...

**Comment by louisblankemeier** [comment]
Yes, if there are missing slices or nonuniform spacing, this could cause issues. I'll add a dataset integrity check to our todos. 

The T12 - L5 vertebral levels need to be visible in CTs that are processed with C2C as we do muscle and adipose tissue analysis at those levels. One todo for us is allowing users to have a subset of these vertebral levels visible and still have the pipeline run. For now, we assume that is there are less than 300 slices, probably not all of the T12 - L5 levels are visible. 

Additionally, our current muscle and adipose tissue model was trained at the L3 level so generalizes worse as you move away from L3. We are currently working on a model that is trained at T12 - L5 and this should be out soon.

**Comment by louisblankemeier** [comment]
Hi Felix,
I just pushed a change that should fix the issue when there is only one connected component on the sagittal slice. Let me know if this helps!

**Comment by louisblankemeier** [comment]
Hi Felix, 

Reopening this issue as I think this is great feedback and we should be able to make it work well for you. Thank you again for all the feedback. To address each of your thoughts:

- I'm glad the issue with patients with scoliosis was fixed. I didn't think about how this could happen when the spine is severely rotated. 
- Yeah this makes sense. I'd also pull the most recent version of Comp2Comp and reinstall Comp2Comp. We found a bug with the orientation and HU values of scans being passed into the segmentation model. The segmentation results should be slightly better now. 
- I agree with you that the 300 slices requirement is not good. I will work on a change that allows you to process any size CT volume and just computes values with any visible T12 - L5 spine level. I'll try to get this in within the next few days. 

Thanks for hanging in there with us as we iron out these issues.

Best regards,
Louis

**Comment by louisblankemeier** [comment]
Hi Felix,

I apologize for the delay. I was close to pushing a solution and then we had a large power outage here. Will push the solution when are servers are back up.

Best regards,
Louis

**Comment by louisblankemeier** [comment]
Hi Felix,

Thank you so much for your patience and sorry for the delay! You should now be able to run the spine_muscle_adipose_tissue pipeline on a series with any number of slices. However, I've kept a lower threshold of 30 because I don't think series with less than 30 slices would be useful in most cases. Feel free to play around with this though.

Please let me know if this works better for you and always grateful for any additional feedback you have.

Thank you,
Louis

**Comment by fohofmann** [comment]
Hi Louis,

no worries at all, seriously! Once again, thank you for all your efforts.

I tested your latest commits and it worked fine (although it might also be due to the fact that we cleaned the dataset more consistently 🙂). The 30 slices requirement seems to be reasonable for me too, especially as the spine segmentation becomes less reliable with less data in the z-axis.

I am going to update the Mac arm64 instructions in the next days as your pull of the more recent TotalSegmentator made some adaptions irrelevant.

Thank you again,
Best, Felix
